lib/ladybug/energy_model/space_type.rb

```diff
@@ -0,0 +1,127 @@
+require 'ladybug/energy_model/gas_equipment_abridged'
+require 'ladybug/energy_model/infiltration_abridged'
+require 'ladybug/energy_model/ventilation_abridged'
+require 'ladybug/energy_model/setpoint_thermostat'
+require 'ladybug/energy_model/setpoint_humidistat'
+
+require 'json-schema'
+require 'json'
+require 'openstudio'
+
+module Ladybug
+  module EnergyModel
+    class SpaceType < ModelObject
+      attr_reader :errors, :warnings
+  
+      def initialize(hash = {})
+        super(hash)
+    
+        raise "Incorrect model type '#{@type}'" unless @type == 'ProgramTypeAbridged'
+      end
+    
+      def defaults
+        result = {}
+        result
+      end
+    
+      def find_existing_openstudio_object(openstudio_model)
+        model_space_type = openstudio_model.getSpaceTypeByName(@hash[:name])
+        return model_space_type.get unless model_space_type.empty?
+        nil
+      end
+    
+      def create_openstudio_object(openstudio_model)    
+        openstudio_space_type = OpenStudio::Model::SpaceType.new(openstudio_model)
+        openstudio_space_type.setName(@hash[:name])
+
+        if @hash[:people]
+          people = PeopleAbridged.new(@hash[:people])
+          openstudio_people = people.to_openstudio(openstudio_model)
+          openstudio_people.setSpaceType(openstudio_space_type)
+        end
+
+        if @hash[:lighting]
+          lights = LightingAbridged.new(@hash[:lighting])
+          openstudio_lights = lights.to_openstudio(openstudio_model)
+          openstudio_lights.setSpaceType(openstudio_space_type)
+        end
+
+        if @hash[:electrical_equipment]
+          electrical_equipment = ElectricalEquipmentAbridged.new(@hash[:electrical_equipment])
+          openstudio_electrical_equipment = electrical_equipment.to_openstudio(openstudio_model)
+          openstudio_electrical_equipment.setSpaceType(openstudio_space_type)
+        end
+
+        if @hash[:gas_equipment]
+         gas_equipment = GasEquipmentAbridged.new(@hash[:gas_equipment])
+         openstudio_gas_equipment = gas_equipment.to_openstudio(openstudio_model)
+         openstudio_gas_equipment.setSpaceType(openstudio_space_type)
+        end
+         
+        if @hash[:infiltration]
+         infiltration = InfiltrationAbridged.new(@hash[:infiltration])
+         openstudio_infiltration = infiltration.to_openstudio(openstudio_model)
+         openstudio_infiltration.setSpaceType(openstudio_space_type)
+        end
+      
+        if @hash[:ventilation]
+         ventilation = VentilationAbridged.new(@hash[:ventilation])
+         openstudio_ventilation = ventilation.to_openstudio(openstudio_model)
+         openstudio_space_type.setDesignSpecificationOutdoorAir(openstudio_ventilation)
+        end
+
+        if @hash[:setpoint]
+          setpoint_thermostat = SetpointThermostat.new(@hash[:setpoint])
+          openstudio_setpoint_thermostat = setpoint_thermostat.to_openstudio(openstudio_model)
```

Comment:
That is great that you are creating the thermostat but it looks like you never end up assigning a unique thermostat derived from this one to each thermal zone. So, ultimately, we just end up with a thermostat without a parent Zone in the Model, which E+ will throw an exception about.

Ultimately, I think that this code does not belong here because OpenStudio treats thermostats as objects that are unique to each Zone, which is fundamentally different from how the Honeybee schema allows you to create a Setpoint object that can be assigned to a Program Type.

So, rather than having this code here, I think you will need to add a method at the end of the model creation that loops through each Zone and sees if there's an ideal air system on the Zone but no thermostat. If so, you will have to find the hash of the ProgramType that is assigned to the Room, get the Setpoint hash, and use it to create a new unique thermostat for the Zone.